### PR TITLE
Eric DeShetler (2408)

### PR DESCRIPTION
**Pull request by ejdesh**
Completed iteration 3

### CODE REVIEW

File: lib/dmv.rb [hunk 1]
```diff
@@ -9,8 +10,11 @@ def add_facility(facility)
   end
 
   def facilities_offering_service(service)
-    @facilities.find do |facility|
-      facility.services.include?(service)
+    yes = []
```
Comment (nicktu12) [inline lib/dmv.rb:13]: better variable name here, maybe `facility_with_service` ?

File: lib/dmv.rb [hunk 2]
```diff
@@ -9,8 +10,11 @@ def add_facility(facility)
...
+
+    @facilities.each do |facility|
+      yes << facility if facility.services.include?(service)
     end
+    yes.uniq
```
Comment (nicktu12) [inline lib/dmv.rb:18]: Nice use of uniq

File: lib/facility.rb [hunk 3]
```diff
@@ -1,14 +1,62 @@
...
   end
 
-  def add_services(service)
+  def add_service(service)
     @services << service
+    @services = @services.uniq
+  end
+
+  def vehicle_type(vehicle)
+    return :ev if vehicle.electric_vehicle?
```
Comment (nicktu12) [inline lib/facility.rb:24]: Guard statements work here, an ifelse could work too

File: lib/facility.rb [hunk 4]
```diff
@@ -1,14 +1,62 @@
...
+    return :antique if vehicle.antique?
+
+    :regular
+  end
+
+  def collect_fee(vehicle)
```
Comment (nicktu12) [inline lib/facility.rb:30]: Love that your broke this out into its own method - promoting SRP!

File: lib/facility.rb [hunk 5]
```diff
@@ -1,14 +1,62 @@
...
+    fee_schedule = { ev: 200, antique: 25, regular: 100 }
+    fee_due = fee_schedule[vehicle_type(vehicle)]
+
+    @collected_fees += fee_due
+  end
+
+  def register(vehicle)
+    collect_fee(vehicle)
+    @registered_vehicles << vehicle
```
Comment (nicktu12) [inline lib/facility.rb:39]: You could consider making this line into its own, small helper function as well

File: lib/facility.rb [hunk 6]
```diff
@@ -1,14 +1,62 @@
...
+
+    vehicle
+  end
+
+  def administer_written_test(registrant)
+    return false unless @services.include?('Written Test')
+    return false unless registrant.age >= 16 && registrant.permit?
```
Comment (nicktu12) [inline lib/facility.rb:46]: Good conditional chaining here - could probably consolidate lines 45, 46 in this way.

File: lib/facility_factory.rb [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    'CO' => :dmv_office,
+    'NY' => :office_name,
+    'MO' => :name
+  }.freeze
+
+  PHONE_KEYS = {
+    'CO' => :phone,
+    'NY' => :public_phone_number,
+    'MO' => :phone
+  }.freeze
```
Comment (nicktu12) [inline lib/facility_factory.rb:18]: This is a nice way to address the different shapes of your data.

File: lib/registrant.rb [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+  def initialize(name, age, permit = false)
+    @name = name
+    @age = age
+    @permit = permit
+    @license_data = { written: false,
+                      license: false,
+                      renewed: false }
+  end
+
+  def permit?
```
Comment (nicktu12) [inline lib/registrant.rb:15]: Nice job exposing permit data here, and not including it in your attr_reader

File: spec/facility_spec.rb [hunk 9]
```diff
@@ -2,25 +2,115 @@
 
 RSpec.describe Facility do
   before(:each) do
-    @facility = Facility.new({name: 'DMV Tremont Branch', address: '2855 Tremont Place Suite 118 Denver CO 80205', phone: '(720) 865-4600'})
+    @facility_1 = Facility.new({  name: 'DMV Tremont Branch',
```
Comment (nicktu12) [inline spec/facility_spec.rb:5]: Solid setup here!

File: spec/facility_spec.rb [hunk 10]
```diff
@@ -2,25 +2,115 @@
...
+      @facility_1.administer_written_test(@registrant_3)
+      expect(@registrant_3.written?).to eq(false)
+    end
+  end
+
+  describe '#administer_road_test' do
+    before(:each) do
+      @facility_1.add_service('Written Test')
+      @facility_1.add_service('Road Test')
+      @facility_1.administer_written_test(@registrant_1)
```
Comment (nicktu12) [inline spec/facility_spec.rb:95]: Is there another side effect to test after the written test has happened?

File: spec/registrant_spec.rb [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    end
+
+    it 'can earn permit' do
+      expect(@registrant_2.permit?).to eq(false)
+      @registrant_2.earn_permit
+      expect(@registrant_2.permit?).to eq(true)
+    end
+  end
+
+  # 2 check permit?
```
Comment (nicktu12) [inline spec/registrant_spec.rb:30]: Comments? Finish these tests, or delete the comments

File: spec/vehicle_spec.rb [hunk 12]
```diff
@@ -32,5 +32,11 @@
       expect(@bolt.electric_vehicle?).to eq(true)
       expect(@camaro.electric_vehicle?).to eq(false)
     end
+    describe '#get_plate' do
+      it 'only returns a plate for :regular, :ev, or :antique' do
```
Comment (nicktu12) [inline spec/vehicle_spec.rb:36]: Consider testing cruz, bolt here too

File: lib/dmv.rb (hunk 1 above)
Comment (ejdesh) [reply lib/dmv.rb:13]: makes sense

File: lib/facility.rb (hunk 3 above)
Comment (ejdesh) [reply lib/facility.rb:24]: idk why I hate if/else so much. Must be something from my childhood.

**Pull request closed without merging**